File: agenkit-cpp/tests/patterns/test_conversational.cpp
/**
 * @file test_conversational.cpp
 * @brief Tests for Conversational pattern
 */

#include <gtest/gtest.h>
#include "agenkit/patterns/conversational.hpp"
#include <memory>

using namespace agenkit;

// Mock agent for testing
class MockLLMAgent : public core::Agent {
private:
    std::string response_text_;
    int call_count_;

public:
    MockLLMAgent(const std::string& response = "Mock response")
        : response_text_(response), call_count_(0) {}

    std::string name() const override { return "mock_llm"; }

    std::future<core::Result<core::Message, core::AgentError>>
    process(core::Message message) override {
        call_count_++;

        // Check if conversation history is in metadata
        std::string response = response_text_;
        if (message.metadata().contains("conversation_history")) {
            auto history = message.metadata()["conversation_history"];
            response += " (with " + std::to_string(history.size()) + " messages in history)";
        }

        auto msg = core::Message::with_text("assistant", response);
        return core::make_ready_future(
            core::Result<core::Message, core::AgentError>::ok(msg)
        );
    }

    int get_call_count() const { return call_count_; }
};

// Test: Basic conversation
TEST(ConversationalTest, BasicConversation) {
    auto llm = std::make_shared<MockLLMAgent>("Hello!");
    patterns::ConversationalAgent agent(llm);

    auto msg = core::Message::with_text("user", "Hi");
    auto result = agent.process(std::move(msg)).get();

    ASSERT_TRUE(result.is_ok());
    auto response = result.unwrap();

    EXPECT_TRUE(response.content_as_str().find("Hello!") != std::string::npos);
    EXPECT_EQ(agent.get_context_length(), 2); // User message + assistant response
}

// Test: Multi-turn conversation
TEST(ConversationalTest, MultiTurnConversation) {
    auto llm = std::make_shared<MockLLMAgent>("Response");
    patterns::ConversationalAgent agent(llm);

    // Turn 1
    auto msg1 = core::Message::with_text("user", "Message 1");
    agent.process(std::move(msg1)).get();

    // Turn 2
    auto msg2 = core::Message::with_text("user", "Message 2");
    auto result2 = agent.process(std::move(msg2)).get();

    ASSERT_TRUE(result2.is_ok());
    EXPECT_EQ(agent.get_context_length(), 4); // 2 user + 2 assistant messages
}

// Test: System prompt
TEST(ConversationalTest, SystemPrompt) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.system_prompt = "You are a helpful assistant.";

    patterns::ConversationalAgent agent(llm, config);

    // System prompt should be in history
    EXPECT_EQ(agent.get_context_length(), 1);

    auto history = agent.get_history();
    EXPECT_EQ(history[0].role(), "system");
    EXPECT_EQ(history[0].content_as_str(), "You are a helpful assistant.");
}

// Test: History pruning
TEST(ConversationalTest, HistoryPruning) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.max_history = 4; // Keep only 4 messages

    patterns::ConversationalAgent agent(llm, config);

    // Send 5 messages (10 total with responses)
    for (int i = 0; i < 5; i++) {
        auto msg = core::Message::with_text("user", "Message " + std::to_string(i));
        agent.process(std::move(msg)).get();
    }

    // Should be pruned to max_history
    EXPECT_EQ(agent.get_context_length(), 4);

    // Should keep most recent messages
    auto history = agent.get_history();
    EXPECT_TRUE(history.back().content_as_str().find("with 4 messages") != std::string::npos);
}

// Test: System prompt preservation during pruning
TEST(ConversationalTest, SystemPromptPreservation) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.max_history = 5;
    config.system_prompt = "System prompt";
    config.include_system_in_count = false;

    patterns::ConversationalAgent agent(llm, config);

    // Send many messages
    for (int i = 0; i < 10; i++) {
        auto msg = core::Message::with_text("user", "Message " + std::to_string(i));
        agent.process(std::move(msg)).get();
    }

    // System message should be preserved
    auto history = agent.get_history();
    EXPECT_EQ(history[0].role(), "system");
    EXPECT_EQ(history[0].content_as_str(), "System prompt");

    // Total should be system + max_history conversation messages
    EXPECT_EQ(agent.get_context_length(), 6); // 1 system + 5 conversation
}

// Test: Clear history
TEST(ConversationalTest, ClearHistory) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.system_prompt = "System prompt";

    patterns::ConversationalAgent agent(llm, config);

    // Add some messages
    auto msg = core::Message::with_text("user", "Test");
    agent.process(std::move(msg)).get();

    EXPECT_GT(agent.get_context_length(), 1);

    // Clear but keep system
    agent.clear_history(true);
    EXPECT_EQ(agent.get_context_length(), 1);

    auto history = agent.get_history();
    EXPECT_EQ(history[0].role(), "system");
}

// Test: Clear history without keeping system
TEST(ConversationalTest, ClearHistoryNoSystem) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.system_prompt = "System prompt";

    patterns::ConversationalAgent agent(llm, config);

    auto msg = core::Message::with_text("user", "Test");
    agent.process(std::move(msg)).get();

    // Clear without keeping system
    agent.clear_history(false);
    EXPECT_EQ(agent.get_context_length(), 0);
}

// Test: Export/import history
TEST(ConversationalTest, ExportImportHistory) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    // Build some history
    auto msg1 = core::Message::with_text("user", "Message 1");
    agent.process(std::move(msg1)).get();

    auto msg2 = core::Message::with_text("user", "Message 2");
    agent.process(std::move(msg2)).get();

    // Export history
    auto exported = agent.export_history();
    EXPECT_TRUE(exported.is_array());
    EXPECT_EQ(exported.size(), 4); // 2 user + 2 assistant

    // Create new agent and import
    auto llm2 = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent2(llm2);

    agent2.import_history(exported);
    EXPECT_EQ(agent2.get_context_length(), 4);

    // Verify content
    auto history = agent2.get_history();
    EXPECT_EQ(history[0].content_as_str(), "Message 1");
}

// Test: Get/set config
TEST(ConversationalTest, GetSetConfig) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalConfig config;
    config.max_history = 20;
    config.system_prompt = "Test";

    patterns::ConversationalAgent agent(llm, config);

    auto retrieved = agent.get_config();
    EXPECT_EQ(retrieved.max_history, 20);
    EXPECT_TRUE(retrieved.system_prompt.has_value());
    EXPECT_EQ(retrieved.system_prompt.value(), "Test");

    // Update config
    patterns::ConversationalConfig new_config;
    new_config.max_history = 5;
    agent.set_config(new_config);

    EXPECT_EQ(agent.get_config().max_history, 5);
}

// Test: Capabilities
TEST(ConversationalTest, Capabilities) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    auto caps = agent.capabilities();
    EXPECT_EQ(caps.size(), 4);
    EXPECT_EQ(caps[0], "conversation");
    EXPECT_EQ(caps[1], "history");
    EXPECT_EQ(caps[2], "context");
    EXPECT_EQ(caps[3], "multi-turn");
}

// Test: Name
TEST(ConversationalTest, Name) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    EXPECT_EQ(agent.name(), "conversational");
}

// Test: Metadata
TEST(ConversationalTest, Metadata) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    auto msg = core::Message::with_text("user", "Test");
    auto result = agent.process(std::move(msg)).get();

    ASSERT_TRUE(result.is_ok());
    auto response = result.unwrap();

    EXPECT_TRUE(response.metadata().contains("pattern"));
    EXPECT_EQ(response.metadata()["pattern"], "conversational");
    EXPECT_TRUE(response.metadata().contains("history_length"));
    EXPECT_TRUE(response.metadata().contains("turn"));
}

// Test: Null agent error
TEST(ConversationalTest, NullAgentError) {
    EXPECT_THROW(
        patterns::ConversationalAgent(nullptr),
        std::invalid_argument
    );
}

// Test: Invalid import history
TEST(ConversationalTest, InvalidImportHistory) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    // Not an array
    nlohmann::json invalid = nlohmann::json::object();

    EXPECT_THROW(
        agent.import_history(invalid),
        std::invalid_argument
    );
}

// Test: Context message creation
TEST(ConversationalTest, ContextMessageCreation) {
    auto llm = std::make_shared<MockLLMAgent>();
    patterns::ConversationalAgent agent(llm);

    // First message
    auto msg1 = core::Message::with_text("user", "First");
    auto result1 = agent.process(std::move(msg1)).get();
    ASSERT_TRUE(result1.is_ok());

    // Second message should have history context
    // At this point, history has: [user: First, assistant: Mock response]
    auto msg2 = core::Message::with_text("user", "Second");
    auto result2 = agent.process(std::move(msg2)).get();
    ASSERT_TRUE(result2.is_ok());

    auto response = result2.unwrap();
    // After adding "Second", before getting response, history has 3 messages:
    // [user: First, assistant: Mock response, user: Second]
    // But the response will indicate history size when context message was created
    // which includes the user message being processed
    EXPECT_TRUE(response.content_as_str().find("with 3 messages") != std::string::npos);
}

File: tests/patterns/test_conversational.py
"""
Tests for ConversationalAgent pattern.
"""

import pytest

from agenkit import Message
from agenkit.patterns import ConversationalAgent, StreamingConversationalAgent


class MockLLM:
    """Mock LLM client for testing.

    Implements ``complete(messages, **kwargs)`` — the contract in
    ``agenkit.adapters.llm.base.LLM`` that all seven shipped adapters implement.
    This double used to implement ``chat(messages)``, which no adapter has, so it
    kept this file green while ``ConversationalAgent`` was unusable with any real
    LLM (#805).
    """

    def __init__(self, response: str = "Test response"):
        self.response = response
        self.call_count = 0
        self.last_messages = None

    async def complete(self, messages, **kwargs):
        """Generate a mock response."""
        self.call_count += 1
        # Store a copy to avoid reference issues
        self.last_messages = messages.copy()
        return Message(role="assistant", content=self.response)


class MockStreamingLLM(MockLLM):
    """Mock streaming LLM client."""

    async def stream(self, messages, **kwargs):
        """Generate a mock streaming response."""
        self.call_count += 1
        self.last_messages = messages
        # Split response into chunks
        words = self.response.split()
        for word in words:
            yield Message(role="assistant", content=word + " ")


@pytest.mark.asyncio
async def test_conversational_agent_basic():
    """Test basic conversational agent functionality."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    # Send a message
    response = await agent.process(Message(role="user", content="Hello"))

    assert response.content == "Test response"
    assert llm.call_count == 1
    assert agent.get_context_length() == 2  # user + assistant


@pytest.mark.asyncio
async def test_conversational_agent_with_system_prompt():
    """Test agent with system prompt."""
    llm = MockLLM()
    agent = ConversationalAgent(
        llm_client=llm, max_history=10, system_prompt="You are a helpful assistant."
    )

    # System prompt should be in history
    assert agent.get_context_length() == 1
    history = agent.get_history()
    assert history[0].role == "system"
    assert history[0].content == "You are a helpful assistant."

    # Send a message
    await agent.process(Message(role="user", content="Hello"))

    # System prompt + user + assistant
    assert agent.get_context_length() == 3


@pytest.mark.asyncio
async def test_conversational_agent_history_pruning():
    """Test history pruning when limit is reached."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=5)

    # Fill history beyond limit
    for i in range(4):
        await agent.process(Message(role="user", content=f"Message {i}"))

    # Should have pruned to max_history
    assert agent.get_context_length() == 5  # Oldest pruned, keeping most recent
    history = agent.get_history()

    # Most recent messages should be preserved
    assert "Message 3" in history[-2].content


@pytest.mark.asyncio
async def test_conversational_agent_preserves_system_prompt():
    """Test that system prompts are preserved during pruning."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=5, system_prompt="System instruction")

    # Fill history beyond limit
    for i in range(5):
        await agent.process(Message(role="user", content=f"Message {i}"))

    # System prompt should still be there
    history = agent.get_history()
    assert history[0].role == "system"
    assert history[0].content == "System instruction"


@pytest.mark.asyncio
async def test_conversational_agent_clear_history():
    """Test clearing conversation history."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10, system_prompt="System instruction")

    # Add some messages
    await agent.process(Message(role="user", content="Hello"))
    assert agent.get_context_length() > 1

    # Clear history, keeping system prompt
    agent.clear_history(keep_system=True)
    assert agent.get_context_length() == 1
    assert agent.get_history()[0].role == "system"

    # Clear history completely
    agent.clear_history(keep_system=False)
    assert agent.get_context_length() == 0


@pytest.mark.asyncio
async def test_conversational_agent_export_import():
    """Test exporting and importing conversation history."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    # Add some messages
    await agent.process(Message(role="user", content="Hello"))
    await agent.process(Message(role="user", content="How are you?"))

    # Export history
    exported = agent.export_history()
    assert len(exported) == 4  # 2 user + 2 assistant

    # Create new agent and import
    new_agent = ConversationalAgent(llm_client=llm, max_history=10)
    new_agent.import_history(exported)

    assert new_agent.get_context_length() == 4
    assert new_agent.get_history()[0].content == "Hello"


@pytest.mark.asyncio
async def test_conversational_agent_context_passed_to_llm():
    """Test that full context is passed to LLM."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10, system_prompt="System")

    # First message
    await agent.process(Message(role="user", content="First"))
    assert len(llm.last_messages) == 2  # system + user

    # Second message
    await agent.process(Message(role="user", content="Second"))
    assert len(llm.last_messages) == 4  # system + user + assistant + user


@pytest.mark.asyncio
async def test_conversational_agent_no_system_prompt():
    """Test agent without system prompt."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    assert agent.get_context_length() == 0

    await agent.process(Message(role="user", content="Hello"))
    assert agent.get_context_length() == 2


@pytest.mark.asyncio
async def test_conversational_agent_system_prompt_not_included():
    """Test agent with system prompt but not included in history."""
    llm = MockLLM()
    agent = ConversationalAgent(
        llm_client=llm,
        max_history=10,
        system_prompt="System instruction",
        include_system=False,
    )

    # System prompt should not be in history
    assert agent.get_context_length() == 0

    await agent.process(Message(role="user", content="Hello"))
    assert agent.get_context_length() == 2


@pytest.mark.asyncio
async def test_streaming_conversational_agent():
    """Test streaming conversational agent."""
    llm = MockStreamingLLM(response="Hello world test")
    agent = StreamingConversationalAgent(llm_client=llm, max_history=10)

    # Stream a response
    chunks = []
    async for chunk in agent.stream(Message(role="user", content="Test")):
        chunks.append(chunk.content)

    # Should have streamed 3 words
    assert len(chunks) == 3
    assert "".join(chunks).strip() == "Hello world test"

    # Full response should be in history
    history = agent.get_history()
    assert len(history) == 2  # user + assistant
    assert history[1].content.strip() == "Hello world test"


@pytest.mark.asyncio
async def test_streaming_agent_preserves_context():
    """Test that streaming agent preserves context across turns."""
    llm = MockStreamingLLM(response="Response")
    agent = StreamingConversationalAgent(llm_client=llm, max_history=10)

    # First turn
    async for _ in agent.stream(Message(role="user", content="First")):
        pass

    # Second turn
    async for _ in agent.stream(Message(role="user", content="Second")):
        pass

    # Should have both turns in history
    assert agent.get_context_length() == 4
    history = agent.get_history()
    assert history[0].content == "First"
    assert history[2].content == "Second"


@pytest.mark.asyncio
async def test_conversational_agent_with_metadata():
    """Test that message metadata is preserved."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    # Add message with metadata
    msg = Message(role="user", content="Hello", metadata={"user_id": "123"})
    await agent.process(msg)

    # Export and check metadata
    exported = agent.export_history()
    assert exported[0]["metadata"]["user_id"] == "123"


@pytest.mark.asyncio
async def test_conversational_agent_edge_case_max_history_zero():
    """Test edge case with max_history=0."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=0)

    await agent.process(Message(role="user", content="Hello"))

    # Should still prune to 0
    assert agent.get_context_length() == 0


@pytest.mark.asyncio
async def test_conversational_agent_edge_case_max_history_one():
    """Test edge case with max_history=1."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=1)

    await agent.process(Message(role="user", content="Hello"))

    # Should keep only 1 message (most recent)
    assert agent.get_context_length() == 1


@pytest.mark.asyncio
async def test_conversational_agent_name_property():
    """Test that agent has a name property."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    assert agent.name == "ConversationalAgent"


@pytest.mark.asyncio
async def test_get_history_returns_copy():
    """Test that get_history returns a copy, not reference."""
    llm = MockLLM()
    agent = ConversationalAgent(llm_client=llm, max_history=10)

    await agent.process(Message(role="user", content="Hello"))

    history = agent.get_history()
    original_length = len(history)

    # Modify the returned history
    history.append(Message(role="user", content="Extra"))

    # Agent's history should not be affected
    assert agent.get_context_length() == original_length
